### ISSUE TITLE
[artifactory-ha] NFS example doesn't work as directed

### ISSUE THREAD

**Issue by TJM** [jfrog/charts#688]
**Is this a request for help?**: Yes (also filed a ticket with jfrog support)

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): bug report

**Version of Helm and Kubernetes**:
helm:

```
version.BuildInfo{Version:"v3.1.1", GitCommit:"afe70585407b420d0097d07b21c47dc511525ac8", GitTreeState:"clean", GoVersion:"go1.13.8"}
```

kubernetes:

```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-13T18:06:54Z", GoVersion:"go1.13.8", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.2", GitCommit:"f6278300bebbb750328ac16ee6dd3aa7d3549568", GitTreeState:"clean", BuildDate:"2019-08-05T09:15:22Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Which chart**: artifactory-ha-2.0.29 

**What happened**: I set the following, in addition to the values-large.yaml and licenses.
```
artifactory:
  persistence:
    type: nfs
    nfs:
      ip: 10.9.35.100
      haDataMount: /artifactory-poc2/data
      haBackupMount: /artifactory-poc2/backup
```

After uploading an artifact, I noticed that the NFS directory was still empty, and after some further investigation, I found that it was not using the mounted NFS directory, instead it was using a persistent RWO pvc.

**What you expected to happen**:
I expected artifactory to use the NFS mount for storing artifacts. It should maybe *not* use a persistent volume (or maybe we should also make it significantly smaller if it *needs* a persistent volume). However it should certainly store the artifacts in the NFS mounted volume.

**How to reproduce it** (as minimally and precisely as possible):
Follow the NFS example from Readme.md... in our case, the NFS server did not create two exports, rather it was under a directory. so we mounted it elsewhere, and created the directories (that feels like prime opportunity for an init container). SO we had to add the yaml indicated above to the values-large.yaml along with our licenses. then...

helm upgrade --install poc2 jfrog/artifactory-ha -f values-nfs.yaml

... then login to the resulting service and upload an artifact to a generic-local repo and observe the results.

**Anything else we need to know**:

As per below, it did not replicate the artifact to the other nodes, so it at least partially was setup to use NFS. I feel like *maybe* the issue is the default mountpoint (`      dataDir: "/var/opt/jfrog/artifactory-ha"`) in values.yaml, that should probably match the default place that the filesystem provider is planning to put data, since the resulting binaryStore.xml was:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!--
  ~ Artifactory is a binaries repository manager.
  ~ Copyright (C) 2019 JFrog Ltd.
  ~
  ~ Artifactory is free software: you can redistribute it and/or modify
  ~ it under the terms of the GNU Affero General Public License as published by
  ~ the Free Software Foundation, either version 3 of the License, or
  ~ (at your option) any later version.
  ~
  ~ Artifactory is distributed in the hope that it will be useful,
  ~ but WITHOUT ANY WARRANTY; without even the implied warranty of
  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  ~ GNU Affero General Public License for more details.
  ~
  ~ You should have received a copy of the GNU Affero General Public License
  ~ along with Artifactory.  If not, see <http://www.gnu.org/licenses/>.
  -->
<!--
    This file configures Artifactory's binary provider
-->

<!--
  ~ TAKE CARE! Making changes to this file may result in losing binaries stored in Artifactory!
  ~ If you are not sure of what you are doing, please contact JFrog Support for assistance.
  -->

<config version="1">
    <chain template="file-system"/>
</config>
```

```
$ k exec -it poc2-artifactory-ha-primary-0 -c artifactory-ha -- df -h | grep artifactory
tmpfs                    31.4G      8.0K     31.4G   0% /artifactory_bootstrap/artifactory.cluster.license
                         50.0G    109.5M     49.9G   0% /artifactory_bootstrap/info/installer-info.json
10.9.35.100:/artifactory-poc2/backup
                        100.0G         0    100.0G   0% /var/opt/jfrog/artifactory-backup
/dev/sds                199.9G     35.5M    199.9G   0% /var/opt/jfrog/artifactory
10.9.35.100:/artifactory-poc2/data
                        100.0G         0    100.0G   0% /var/opt/jfrog/artifactory-ha
```

```
$ k exec -it poc2-artifactory-ha-member-0 -c artifactory-ha -- df -h | grep artifactory
                         50.0G     45.9M     49.9G   0% /artifactory_bootstrap/info/installer-info.json
/dev/sdk                199.9G     36.1M    199.9G   0% /var/opt/jfrog/artifactory
10.9.35.100:/artifactory-poc2/data
                        100.0G         0    100.0G   0% /var/opt/jfrog/artifactory-ha
10.9.35.100:/artifactory-poc2/backup
                        100.0G         0    100.0G   0% /var/opt/jfrog/artifactory-backup
```

```
$ k exec -it poc2-artifactory-ha-member-1 -c artifactory-ha -- df -h | grep artifactory
                         50.0G     51.2M     49.9G   0% /artifactory_bootstrap/info/installer-info.json
10.9.35.100:/artifactory-poc2/backup
                        100.0G         0    100.0G   0% /var/opt/jfrog/artifactory-backup
/dev/sdf                199.9G    100.0M    199.8G   0% /var/opt/jfrog/artifactory
10.9.35.100:/artifactory-poc2/data
                        100.0G         0    100.0G   0% /var/opt/jfrog/artifactory-ha
```

```
$ k get pvc -l release=poc2
NAME                                   STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
data-poc2-postgresql-0                 Bound    pvc-5ef27ae7-69a7-4525-9744-90022c8ea46c   50Gi       RWO            vsan-default   37m
poc2-artifactory-ha-backup-pvc         Bound    poc2-artifactory-ha-backup-pv              200Gi      RWO                           37m
poc2-artifactory-ha-data-pvc           Bound    poc2-artifactory-ha-data-pv                200Gi      RWO                           37m
volume-poc2-artifactory-ha-member-0    Bound    pvc-5af00a3b-f7f5-44ed-a8f5-78291116b4fb   200Gi      RWO            vsan-default   37m
volume-poc2-artifactory-ha-member-1    Bound    pvc-0dc8d1a4-0e6d-44da-a2ba-03a24c722c1d   200Gi      RWO            vsan-default   33m
volume-poc2-artifactory-ha-primary-0   Bound    pvc-686ee72e-4544-4ff8-8af4-b37445b06bd5   200Gi      RWO            vsan-default   37m
```

**Comment by eldada** [comment]
Right... seems we have a bug here 😬 
The servers configuration is now controlled by the `system.yaml`, which we manage with the `artifactory.systemYaml` key.
The key that triggers the use of the NFS in the `system.yaml` is 
```
artifactory:
...
  systemYaml: |
    ...
    ...
    artifactory:
      node:
        haDataDir: /artifactory-poc2/data
```

So in your case, to get it working, you need to set in your `values-poc.yaml` something like
```
artifactory:
  ...
  # This directory is intended for use with NFS eventual configuration for HA
  haDataDir:
    enabled: true
    path: /artifactory-poc2/data

```

Can you test this and let us know if it worked?

**Comment by TJM** [comment]
I went with:

```
artifactory:
  persistence:
    type: nfs
    nfs:
      ip: 10.9.35.100
      haDataMount: /artifactory-poc2/data
      haBackupMount: /artifactory-poc2/backup
  haDataDir:
      enabled: true
      path: "{{ .Values.artifactory.persistence.nfs.dataDir }}"
```

I also had to set:

```
postgresql:
  postgresqlPassword: (retrieved from current secret)
```

... and it looked like it was going to work, but after the `helm upgrade --install poc2 jfrog/artifactory-ha -f values-nfs.yaml`, the artifactory pods got stuck waiting for the primary. As this is just a POC, I can just delete it and start over, but it feels like there might be an order of operations issue too. (will dig into that in another issue)

Upon trying to delete and start over, I found that it would not even start up... waiting on database. It seems that it does not delete its PVCs during `helm delete NAME` (is that on purpose?), *and* it uses predictable names for the PVCs (probably bad?) so it ends up re-using the same PV's which apparently have some old/broken configuration still :(

**Comment by eldada** [comment]
By design, when you `helm delete` a release, all the StatefulSets leave their PVCs behind to protect against data loss. This is built into the StatefulSet object.
You have to manually delete PVCs with an explicit `kubectl delete pvc ...`
We are working on fixing the docs and examples. Hope this gets resolved in the coming days.

**Comment by TJM** [comment]
OK, so setting haDataDir settings seemed to work...

```
$ for pod in $(kubectl get pod --no-headers | awk '/artifactory/ {print $1}'); do echo "*** $pod ***"; kubectl exec -it $pod -c artifactory-ha -- df -h | grep jfrog; done
*** poc1-artifactory-ha-member-0 ***
/dev/sde                199.9G    169.3M    199.7G   0% /var/opt/jfrog/artifactory
*** poc1-artifactory-ha-member-1 ***
/dev/sdf                199.9G    168.7M    199.7G   0% /var/opt/jfrog/artifactory
*** poc1-artifactory-ha-primary-0 ***
/dev/sde                199.9G     42.4M    199.9G   0% /var/opt/jfrog/artifactory
*** poc2-artifactory-ha-member-0 ***
                        100.0G     64.5M     99.9G   0% /var/opt/jfrog/artifactory-backup
/dev/sdi                199.9G     35.7M    199.9G   0% /var/opt/jfrog/artifactory
                        100.0G     64.5M     99.9G   0% /var/opt/jfrog/artifactory-ha
*** poc2-artifactory-ha-member-1 ***
                        100.0G     64.5M     99.9G   0% /var/opt/jfrog/artifactory-backup
/dev/sdg                199.9G     35.5M    199.9G   0% /var/opt/jfrog/artifactory
                        100.0G     64.5M     99.9G   0% /var/opt/jfrog/artifactory-ha
*** poc2-artifactory-ha-primary-0 ***
/dev/sdk                199.9G     35.5M    199.9G   0% /var/opt/jfrog/artifactory
                        100.0G     64.5M     99.9G   0% /var/opt/jfrog/artifactory-ha
                        100.0G     64.5M     99.9G   0% /var/opt/jfrog/artifactory-backup
```
(poc1 - filesystem, poc2 - nfs)

Is there a way to programmatically set the haDataDir settings, like if the persistence type is nfs, then set enabled to true and set the path to the mountpoint? 

Also, the manual operation of creating the directories seems like an anti-pattern...
* Is there a way to "skip" the backup mount, and just mount the haDataMount directly (not mounting a sub-directory)? NOTE: Our NFS datastore is currently around 5TB/8TB, but the backup is over 12TB. The backup mechanism seems to fall on its face with that size of datastore.
* Most usefully, if we were to keep the backup mount, it would most likely be a separate NFS server IP and mountpoint, not just another sub-directory on the same server/mountpoint. What would be the point of backing up to the same array? :)

~tommy

**Comment by TJM** [comment]
> By design, when you `helm delete` a release, all the StatefulSets leave their PVCs behind to protect against data loss. This is built into the StatefulSet object.
> You have to manually delete PVCs with an explicit `kubectl delete pvc ...`
> We are working on fixing the docs and examples. Hope this gets resolved in the coming days.

Thanks! I did delete the pvc's, and I am not sure exactly how to document that situation, maybe an uninstall or re-install section? Thanks for clarifying why this behaved the way it does, and it seems somewhat reasonable to keep the PVCs, but I sorta wish helm delete had a --purge  ;)

**Comment by eldada** [comment]
Should now be fixed with #697.

**Comment by TJM** [comment]
@eldada I still need some help on this. I appreciate the separation of these options, but that is only a partial solution. Should I re-open this and continue or open another issue with the specific questions?

**Comment by eldada** [comment]
Yes. Let's continue there.
